### PR TITLE
Add support for PKCS#11 3.2 validation objects

### PR DESCRIPTION
**Pull request by Jakuje**
The PKCS#11 3.2 introduces a way to query the session for the validation flags of the last operation. This is done with the new API `C_GetSessionValidationFlags` which is being exposed now also to the users of this crate.

It also defines the new attribute specifying if the given object is matching the requirements for the validation.

Last but not least there is new validation object exposing information about the validation itself.

There is currenly almost no coverage for these, as this is implemented only by kryoptic (as far as I know) and not enabled in the default build we are using in CI.

Opening as draft as it depends on some fixes in kryoptic (latchset/kryoptic#315) as well its based on other code changes here (#304).

### CODE REVIEW

File: cryptoki/src/session/validation.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    val: CK_SESSION_VALIDATION_FLAGS_TYPE,
+}
+
+impl ValidationFlagsType {
+    /// Check the last operation met all requirements of a validated mechanism.
+    pub const VALIDATION_OK: ValidationFlagsType = ValidationFlagsType {
+        val: CKS_LAST_VALIDATION_OK,
+    };
+
+    pub(crate) fn stringify(flags: CK_SESSION_VALIDATION_FLAGS_TYPE) -> String {
```
Comment (wiktor-k) [inline cryptoki/src/session/validation.rs]: I think this can be inlined into `Display`, like so: https://github.com/parallaxsecond/rust-cryptoki/pull/304/files#diff-8eed47b951056624bcbe744076eac12fab93c4fe6c655c937351243a0db816daR1775

File: cryptoki/src/session/validation.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    }
+}
+
+impl std::fmt::Display for ValidationFlagsType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", ValidationFlagsType::stringify(self.val))
+    }
+}
+
+impl Deref for ValidationFlagsType {
```
Comment (wiktor-k) [inline cryptoki/src/session/validation.rs]: `AsRef` would be safer IMO.

File: cryptoki/tests/basic.rs [hunk 3]
```diff
@@ -3822,13 +3821,87 @@ fn unique_id() -> TestResult {
...
+    // generate a secret key
+    let key = session.generate_key(&Mechanism::AesKeyGen, &generate_template)?;
+
+    // we can get the ObjectValidationFlags attribute
+    let attrs = session.get_attributes(key, &[AttributeType::ObjectValidationFlags])?;
+    if is_softhsm() {
+        // SoftHSM does not support this attribute at all
+        assert_eq!(attrs.len(), 0);
+    } else {
+        // Kryoptic supports the ObjectValidationFlag only if it is built as a FIPS provider
```
Comment (wiktor-k) [inline cryptoki/tests/basic.rs:4265]: :thinking: is it possible to built it this way for our CI? I think it'd be cool to have some real results here :)

Comment (Jakuje) [reply cryptoki/tests/basic.rs:4265]: Sure. But its slow as it builds also the whole OpenSSL :) I was trying this locally and there are some more limitations and corner cases, which do not work or work differently and which we would have to workaround. But with the validation API, it should be quite easy to distinguish the FIPS and non-FIPS kryoptic version.

File: cryptoki/src/session/validation.rs (hunk 2 above)
Comment (Jakuje) [reply cryptoki/src/session/validation.rs]: Resolved.

File: cryptoki/src/session/validation.rs (hunk 1 above)
Comment (Jakuje) [reply cryptoki/src/session/validation.rs]: Agreed.  Done.

File: cryptoki/tests/basic.rs (hunk 3 above)
Comment (Jakuje) [reply cryptoki/tests/basic.rs:4265]: Added the FIPS build of kryoptic and adjusted the tests to work with it. The build is slow, but after OpenSSL build will be cached, it should be faster (but still slower than other tests).

File: cryptoki-sys/build.rs [hunk 4]
```diff
@@ -31,6 +31,7 @@ mod generate {
                 ("CKC_", "CK_CERTIFICATE_TYPE"),
                 ("CKD_", "CK_EC_KDF_TYPE"),
                 ("CKF_", "CK_FLAGS"),
+                ("CKV_", "CK_ULONG"),
```
Comment (hug-dev) [inline cryptoki-sys/build.rs:34]: Good catch!

File: cryptoki/src/session/validation.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    fn from(val: ValidationFlagsType) -> Self {
+        *val.as_ref()
+    }
+}
+
+impl Session {
+    /// Get requested validation flags from the session
+    ///
+    /// The only supported flag as for PKCS#11 3.2 is `ValidationFlagsType::VALIDATION_OK`
+    pub fn get_validation_flags(&self, flags_type: ValidationFlagsType) -> Result<CK_FLAGS> {
```
Comment (hug-dev) [inline cryptoki/src/session/validation.rs:53]: Could we use any of the existing abstraction for the `CK_FLAGS` returned here?

File: .github/workflows/kryoptic-fips.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
+---
+name: Test kryoptic FIPS module
```
Comment (hug-dev) [inline .github/workflows/kryoptic-fips.yml:2]: Shall I add it to the required workflow to pass for a PR?

**Review by hug-dev — DISMISSED** [review]
Thank you!

File: .github/workflows/kryoptic-fips.yml (hunk 6 above)
Comment (wiktor-k) [reply .github/workflows/kryoptic-fips.yml:2]: I think now this can be done using the "[Check if all checks succeeded (pull_request)](https://github.com/parallaxsecond/rust-cryptoki/actions/runs/17860753123/job/50880857592?pr=306)" job, thus, Jakub can adjust it himself.

File: cryptoki/src/session/validation.rs (hunk 5 above)
Comment (Jakuje) [reply cryptoki/src/session/validation.rs:53]: The resulting `CK_FLAGS` value here are vendor specific that are based on the  validation object (`CKA_VALIDATION_FLAG` attribute). That said, we would have to define the ValidationFlags abstraction, but the user would have to define its own vendor specific values. I did not work with the vendor specific atrributes in rust-cryptoki so far, but is probably the cleanest approach. It will also make it more clear that these two flags should be related to each other.

https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.2/pkcs11-spec-v3.2.html#_Toc195693109

Comment (Jakuje) [reply cryptoki/src/session/validation.rs:53]: ugh ... not sure if we can easily make a vendor defined bitfield though ...

Comment (hug-dev) [reply cryptoki/src/session/validation.rs:53]: I see.. we can probably keep it like that for now and find a good abstraction when it's getting used more?

File: .github/workflows/kryoptic-fips.yml (hunk 6 above)
Comment (Jakuje) [reply .github/workflows/kryoptic-fips.yml:2]: I am not sure if it would not need to be in the same workflow file to work though to be able to use in the "check if all checks succeeded". Given that how large this workflow is, I did not want to mess the main ci.yml with it for now.

I hope this job will be stable, but I would rather keep it for some time non-mandatory and add it to the required just after some time we will see it will work as expected to avoid working around some required jobs.

File: cryptoki/src/session/validation.rs (hunk 5 above)
Comment (Jakuje) [reply cryptoki/src/session/validation.rs:53]: Yeah, that would be my preference as I am not sure where to start defining the vendor specific bit field.

**Review by hug-dev — DISMISSED** [review]

File: .github/workflows/kryoptic-fips.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      #################
+      ### the tests ###
+      #################
+      - name: Run test script
+        env:
+          KRYOPTIC_CONF: /tmp/kryoptic.sql
+          TEST_PKCS11_MODULE: /__w/rust-cryptoki/kryoptic/target/debug/libkryoptic_pkcs11.so
+          OUT_DIR: /__w/rust-cryptoki/kryoptic/target/debug/deps/
+        run: |
+          RUST_BACKTRACE=1 cargo build --all-features &&
```
Comment (wiktor-k) [inline .github/workflows/kryoptic-fips.yml]: I think `RUST_BACKTRACE=1` can also be moved to `env`.
I don't think we need `cargo build` here... `cargo check` should be sufficient :thinking:

File: cryptoki/tests/basic.rs [hunk 8]
```diff
@@ -2378,7 +2484,7 @@ fn kbkdf_feedback_mode() -> TestResult {
     let dkm_length_format = KbkdfDkmLengthFormat::new(
         KbkdfDkmLengthMethod::SumOfKeys,
```
Comment (wiktor-k) [inline cryptoki/tests/basic.rs:2768]: Could you explain why `SumOfKeys` is OK here but in the fragment above it was changed to `SumOfSegments`? :thinking: Thanks!

Comment (Jakuje) [reply cryptoki/tests/basic.rs:2768]: I did not dig deep enough, but generally, kryoptic implements all possible PKCS#11 combinations in Rust. But for the FIPS we want to use the OpenSSL implementation, which has limitation that the Counter mode supports only sum of segments, while Feedback mode supports only the sum of keys.

I reverted the changes to the double-pipeline KDFs as they are untested (I think not implemented in OpenSSL) and I am not sure how they will work.

File: .github/workflows/kryoptic-fips.yml (hunk 7 above)
Comment (Jakuje) [reply .github/workflows/kryoptic-fips.yml]: Moved the env + removed the separate build

**Review by wiktor-k — APPROVED** [review]
LGTM :+1: thanks!

**Comment by Jakuje** [comment]
Thanks! That was fast. Lest hope the CI with FIPS kryoptic will work :) Otherwise expect follow-up PR :)
